Src/ILGPU/Context.cs

```diff
@@ -384,6 +385,70 @@ public DeviceCollection<TDevice> GetDevices<TDevice>()
                 : new DeviceCollection<TDevice>(new List<Device>());
         }
 
+        /// <summary>
+        /// Attempts to return the most optimal single device.
+        /// </summary>
+        /// <param name="preferCPU">Always returns CPU device 0.</param>
+        /// <returns>Selected device.</returns>
+        public Device GetPreferredDevice(bool preferCPU)
+        {
+            if (preferCPU)
+            {
+                return GetDevice<CPUDevice>(0);
+            }
+
+            var sorted = Devices
+                .OrderByDescending(d => d.MemorySize)
+                .Where(d => d.AcceleratorType != AcceleratorType.CPU);
+
+            return sorted.FirstOrDefault() ?? GetDevice<CPUDevice>(0);
+        }
+
+        /// <summary>
+        /// Attempts to return the most optimal set of devices.
+        /// </summary>
+        /// <param name="preferCPU">Always returns first CPU device.</param>
+        /// <param name="matchingDevicesOnly">Only returns matching devices.</param>
+        /// <returns>Selected devices.</returns>
+        public IEnumerable<Device> GetPreferredDevices(bool preferCPU, bool matchingDevicesOnly)
+        {
+            if (preferCPU)
+            {
+                return deviceMapping.TryGetValue(AcceleratorType.CPU, out var devices)
+                    ? devices
+                    : throw new NotSupportedException(
+                            RuntimeErrorMessages.NotSupportedTargetAccelerator);
+            }
+
+            var sorted = Devices
+                .OrderByDescending(d => d.MemorySize)
+                .Where(d => d.AcceleratorType != AcceleratorType.CPU);
+
+            if (sorted.Any())
+            {
+                if (matchingDevicesOnly)
+                {
+                    Device toMatch = sorted.First();
+
+                    return sorted.Where(
+                        d =>
+                        d.AcceleratorType == toMatch.AcceleratorType &&
+                        d.MemorySize == toMatch.MemorySize);
+                }
+                else
+                {
+                    return sorted;
+                }
+            }
+            else
+            {
+                return deviceMapping.TryGetValue(AcceleratorType.CPU, out var devices)
```

Comment:
This looks like GetDevice<CPUDevice>(0) too.

Reply:
Same here... :nerd

Reply:
The PR is correct - `GetCPUDevices` returns `DeviceCollection`, which does not implement `IEnumerable<>T`.